beluga/include/beluga/estimation/cluster_based_estimator.hpp

```diff
@@ -0,0 +1,372 @@
+      return (
+          (it != grid_cell_data.end()) &&            // is within the map
+          (!it->second.cluster_id.has_value()) &&    // AND not yet mapped to a cluster
+          (it->second.weight <= grid_cell_weight));  // AND has lower weight than the current cell
+    };
+
+    auto valid_neighbor_hashes_view =                     //
+        neighbors |                                       //
+        ranges::views::transform(get_neighbor_hash) |     //
+        ranges::views::filter(filter_invalid_neighbors);  //
+
+    for (const auto& neighbor_hash : valid_neighbor_hashes_view) {
+      auto& neighbor = grid_cell_data[neighbor_hash];
+      neighbor.cluster_id = grid_cell.cluster_id;
+      const auto inflated_priority =
+          weight_cap + neighbor.weight;  // since weights are capped at weight_cap, this gives us a value that is
+                                         // guaranteed to be higher than any other weight from a local maximum.
+      grid_cell_queue.push({inflated_priority, neighbor_hash});  // reintroduce with inflated priority
+    }
+  }
+}
+
+/// @brief For each cluster, estimate the mean and covariance of the states that belong to it.
+/// @tparam StateType Type used for the pose estimation.
+/// @tparam CovarianceType Type used for the covariance estimation.
+/// @tparam GridCellDataMapType Type of the grid cell data map.
+/// @tparam Range Range type of the states.
+/// @tparam Weights Range type of the weights.
+/// @tparam Hashes Range type of the hashes.
+/// @param grid_cell_data Grid cell data map that has already been processed by the clustering algorithm.
+/// @param states Range containing the states of the particles.
+/// @param weights Range containing the weights of the particles.
+/// @param hashes Range containing the hashes of the particles.
+/// @return A vector of tuples, containing the weight, mean and covariance of each cluster, in no particular order.
+template <class StateType, class CovarianceType, class GridCellDataMapType, class Range, class Weights, class Hashes>
+[[nodiscard]] auto
+estimate_clusters(GridCellDataMapType&& grid_cell_data, Range&& states, Weights&& weights, Hashes&& hashes) {
+  struct ClusterInfo {
+    double total_weight{0.0};
+    std::size_t particle_count{0};
+  };
+  std::unordered_map<std::size_t, ClusterInfo> cluster_info;
+
+  // find out the weight of each of the clusters, along with how many clusters there are
+  for (const auto& [weight, hash] : ranges::views::zip(weights, hashes)) {
+    const auto& grid_cell = grid_cell_data[hash];
+    if (!grid_cell.cluster_id.has_value()) {
+      continue;
+    }
+    auto [it, inserted] = cluster_info.try_emplace(*grid_cell.cluster_id, ClusterInfo{});
+    it->second.total_weight += weight;
+    ++it->second.particle_count;
+  }
+
+  // for each cluster found, estimate the mean and covariance of the states that belong to it
+  std::vector<std::tuple<double, StateType, CovarianceType>> cluster_estimates;
```

Comment:
These types can be deduced instead of being template parameters, which would make this function easier to use outside the mixin context.

```suggestion
  using StateType = std::decay_t<decltype(std::get<0>(beluga::estimate(states, weights)))>;
  using CovarianceType = std::decay_t<decltype(std::get<1>(beluga::estimate(states, weights)))>;
  std::vector<std::tuple<double, StateType, CovarianceType>> cluster_estimates;
```